networkx/algorithms/assortativity/neighbor_degree.py

```diff
@@ -125,12 +133,34 @@ def average_neighbor_degree(G, source="out", target="out", nodes=None, weight=No
     tgt_deg = dict(target_degree())
     # average degree of neighbors
     avg = {}
+    # edges in the graph together with their weights
+    edges = G.edges(data=True)
+
     for n, deg in source_degree(nodes, weight=weight):
         # normalize but not by zero degree
         if deg == 0:
             avg[n] = 0.0
             continue
+
+        # if G is an undirected graph, G_n is neighbors of n
         G_n = G[n]
+
+        """
+        if G is a directed graph, 
+            if source is 'in', G_n will be predecessors of n
+            else if source is 'out', G_n will be successors of n
+        """
+        if G.is_directed():
+            G_n = {}
+            if source == "in":
+                for start, end, weight_dict in edges:
+                    if end == n:
+                        G_n[start] = weight_dict
+            elif source == "out":
+                for start, end, weight_dict in edges:
+                    if start == n:
+                        G_n[end] = weight_dict
```

Comment:
Can this use `G.pred[n]` or `G.succ[n]` instead of `G[n]`?
It feels like we shouldn't have to rebuild the adjacency dict.

Also, how do we handle the "in+out" case? `G[n]` only yields outgoing neighbors. 
So we need a way to chain the incoming and outgoing neighbors, while making sure
to handle a neighbor that is both incoming and outgoing appropriately.

Reply:
If you get an error about not finding `G.succ` for class `Graph`, then you are treating an undirected graph as a directed graph somewhere. Please track that down. Only DiGraph and MultiDiGRaph have the attributes `G.succ` and `G.pred`.

I like your approach of `P_n` and `S_n` along with making them empty dicts when they aren't relevant. 

Two quick comments... Can you use # for comments instead of """?  And is there a way to pull the if statements out of the loop so that logic can be done once before the loop instead. Something like:
```python
    # default to empty neighbors dicts
    G_P = S_P = {n: {} for n in G}
    if G.is_directed():
        if "in" in source:
            G_P = G.pred
        if "out" in source:
            G_S = G.succ
    else:
        G_S = G.adj
```
Then inside the loop you can use `P_n = G_P[n]` and `S_n = G_S[n]`

Reply:
Thank you for your comments. I have updated the code accordingly and committed the new version. If there are any other issues, it would be great to learn and solve them.